### DISCUSSION TITLE
Duplicate variant types, possible?

### DISCUSSION THREAD

**Discussion by Dikymon** [stephenberry/glaze#843] [Q&A] [answered]
Hello

We have a use-case where variant keys can map to the same type and i was hoping to simply extend the put/delete action. eg.

`using tagged_variant = std::variant<put_action, delete_action, delete_action>;

template <>
struct glz::meta<tagged_variant>
{
   static constexpr std::string_view tag = "action";
   static constexpr auto ids = std::array{"PUT", "DELETE", "DELETE2"}; //Defaults to glz::name_v of the type if ids is not supplied
};`

But this errors with

`glaze_example/build/_deps/glaze-src/include/glaze/util/variant.hpp:26:43: error: no match for ‘operator=’ (operand types are ‘std::array<std::variant<put_action, delete_action, delete_action>, 3>::value_type’ {aka ‘std::variant<put_action, delete_action, delete_action>’} and ‘std::variant_alternative_t<2, std::variant<put_action, delete_action, delete_action> >’ {aka ‘delete_action’})
   26 |          for_each<N>([&](auto I) { ret[I] = std::variant_alternative_t<I, T>{}; });
`

Is this simply not possible using the variant method or should we move to a custom handler?

/Me

**Comment by stephenberry** [comment, marked as answer]
It's not a good idea to duplicate a type in a std::variant. So, what you're wanting is two names to refer to the same type. This is currently not supported, as it would add complexity to the code.

However, you can make a strongly typed alias to your type.

Note that `using delete_action2 = delete_action;` does not work because it is still the same type.

To get a different type you need to make a base type that is templated.

```c++
template <int Alias>
struct base_delete_action
{
   // put fields here
};

using delete_action = base_delete_action<1>;

using delete_action2 = base_delete_action<2>;
```

You can use reflection or partial template specialization on `base_delete_action` so that you don't duplicate your `glz::meta`:

```c++
template <int Alias>
struct glz::meta<base_delete_action<Alias>>
{
   // your meta specification here
};
```

Let me know if this solution is acceptable for you. There are other ways to produce strongly typed aliases as well.

Note that with this approach you would make your variant: `std::variant<put_action, delete_action, delete_action2>`
